File: Advance/12.Find the maximum area of a rectangle in a histogram/Solution.py
def maxRectangleArea(heights):
    stack = []
    max_area = 0
    n = len(heights)

    for i in range(n + 1):
        curr_height = heights[i] if i < n else 0
        while stack and curr_height < heights[stack[-1]]:
            h = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, h * width)
        stack.append(i)

    return max_area


if __name__ == "__main__":
    n = int(input("Enter the number of bars in the histogram: "))
    heights = list(map(int, input("Enter the heights of the bars: ").split()))

    result = maxRectangleArea(heights)
    print(f"Maximum area of rectangle: {result}")

File: Advance/12.Find the maximum area of a rectangle in a histogram/Solution.java
import java.util.Scanner;
import java.util.Stack;

public class HistogramMaxArea {
    public static int maxRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0, n = heights.length;

        for (int i = 0; i <= n; i++) {
            int currHeight = (i == n) ? 0 : heights[i];
            while (!stack.isEmpty() && currHeight < heights[stack.peek()]) {
                int h = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, h * width);
            }
            stack.push(i);
        }
        return maxArea;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of bars in the histogram: ");
        int n = sc.nextInt();

        int[] heights = new int[n];
        System.out.print("Enter the heights of the bars: ");
        for (int i = 0; i < n; i++) {
            heights[i] = sc.nextInt();
        }

        int result = maxRectangleArea(heights);
        System.out.println("Maximum area of rectangle: " + result);
    }
}
